README.md
# xlnodeid_py

Initial commit.

## On-line Documentation

More information on the **xlnodeid_py** project can be found
[here](https://jddixon.github.io/xlnodeid_py)


CHANGES
py/xlnodeid_py/CHANGES

v0.0.12
    2018-03-23
        * fix corrupt .dvcz/builds                                  SLOC  584
v0.0.11
    2018-03-09
        * fix setup.py and tox config files                         SLOC  580
v0.0.10
    2017-12-26
        * pylinting; update config files                            SLOC  586
v0.0.9
    2017-10-14
        * add .dvcz/projControl.toml                                SLOC  591
v0.0.8
    2017-08-25
        * add tox.ini, *requirements.txt; plus minor edits          SLOC  550
v0.0.7
    2017-08-04
        * rewrite setup.py; minor edits cfg files                   SLOC  506
v0.0.6
    2017-05-29
        * drop dist/* from repos; plus minor edits                  SLOC  493
v0.0.5
    2017-03-19
        * improved test_xl_node_id.py                               SLOC  490
v0.0.4
    2017-03-18
        * XLNodeID and rudimentary test_xl_node_id.py               SLOC  436
v0.0.3
    2017-03-04
        * add pylint, checkCoverage, test_basics.py                 SLOC  339
v0.0.2
    2017-03-01
        * add configuration files                                   SLOC  200
v0.0.1
    2017-02-28 
        * initial commit















test_basics.py
#!/usr/bin/env python3
# xlxlnodeid_py/test_basics.py

""" Currently just exercises test framework. """

#import hashlib
#import os
import time
import unittest

from rnglib import SimpleRNG
#from xlxlnodeid import *


class TestWhatever(unittest.TestCase):
    """ Currently just exercises test framework. """

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    def test_whatever(self):
        """
        Tests nothing at all.
        """

        self.assertEqual(1, 1)

if __name__ == '__main__':
    unittest.main()


test_xlnodeid.py
#!/usr/bin/env python3
# xlnodeid_py/test_node_id.py

""" Verify that an XLNodeID behaves like one. """

import time
import unittest
from copy import deepcopy

from rnglib import SimpleRNG
from xlattice import SHA1_BIN_LEN, SHA2_BIN_LEN, SHA3_BIN_LEN
from xlnodeid import XLNodeID, XLNodeIDError


class TestNodeID(unittest.TestCase):
    """ Verify that an XLNodeID behaves like one. """

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    def test_valid_node_id(self):
        """
        Tests the funnction of that name.
        """

        # tests that should fail
        self.assertFalse(XLNodeID.is_valid_node_id(None))
        self.assertFalse(XLNodeID.is_valid_node_id('foo'))  # not bytes-like
        self.assertFalse(XLNodeID.is_valid_node_id(b'bar'))  # wrong length
        self.assertFalse(XLNodeID.is_valid_node_id(42))     # an int

        # tests that should succeed
        val = bytes(SHA1_BIN_LEN)
        self.assertTrue(XLNodeID.is_valid_node_id(val))
        val = bytes(SHA2_BIN_LEN)
        self.assertTrue(XLNodeID.is_valid_node_id(val))
        val = bytes(SHA3_BIN_LEN)
        self.assertTrue(XLNodeID.is_valid_node_id(val))

    def expect_failure(self, val):
        """ Expect object construction to fail. """
        try:
            _ = XLNodeID(val)
            self.fail("XLNodeID constructed with bad ID")
        except XLNodeIDError:
            pass

    def expect_success(self, val):
        """
        Try to build a NodeID with val (a bytes valaue).
        """
        try:
            nodeid = XLNodeID(val)
            # succeeded
        except XLNodeIDError:
            self.fail("ctor raised with good ID")

        val2 = nodeid.value
        self.assertIsNotNone(val2)
        self.assertFalse(val2 is val)   # not the same object
        self.assertEqual(val2, val)     # but a valid deep copy

    def test_ctor(self):
        """
        Tests the XLNodeID constructor.
        """

        # tests that should fail
        self.expect_failure(None)    # id may not be None
        self.expect_failure('foo')   # not bytes-like
        self.expect_failure(b'bar')  # wrong length
        self.expect_failure(42)      # an int

        self.expect_failure(bytes(SHA1_BIN_LEN - 1))
        self.expect_failure(bytes(SHA1_BIN_LEN + 1))
        self.expect_failure(bytes(SHA2_BIN_LEN - 1))
        self.expect_failure(bytes(SHA2_BIN_LEN + 1))
        self.expect_failure(bytes(SHA3_BIN_LEN - 1))
        self.expect_failure(bytes(SHA3_BIN_LEN + 1))

        # tests that should succeed
        val = self.rng.some_bytes(SHA1_BIN_LEN)
        self.expect_success(val)
        val = self.rng.some_bytes(SHA2_BIN_LEN)
        self.expect_success(val)
        val = self.rng.some_bytes(SHA3_BIN_LEN)
        self.expect_success(val)

    def do_test_cloning(self, length):
        """ Verify that cloning works for a given number of bytes. """
        val = self.rng.some_bytes(length)
        id1 = XLNodeID(val)
        id2 = id1.clone()
        self.assertTrue(id1 is not id2)
        self.assertEqual(id1.value, id2.value)

    def test_cloning(self):
        """ Test cloning for bytes-like objects of a given number of bytes. """
        for length in [SHA1_BIN_LEN, SHA2_BIN_LEN, SHA3_BIN_LEN]:
            self.do_test_cloning(length)

    def do_test_comparison(self, length):
        """
        For a quasi-random byte sequence of a given length, verify that
        comparison opertors work.
        """
        val = self.rng.some_bytes(length)       # a byte array of that length

        # pick a random index into that byte array
        ndx = 1 + self.rng.next_int16(length - 1)
        if val[ndx] == 0:
            val[ndx] = 1
        if val[ndx] == 255:
            val[ndx] = 254

        # make a couple of clones of the byte array
        v_bigger = deepcopy(val)
        v_bigger[ndx] += 1
        v_smaller = deepcopy(val)
        v_smaller[ndx] -= 1

        self.assertTrue(v_bigger > val)
        self.assertTrue(v_smaller < val)

        # use these values to make NodeIDs
        n_bigger = XLNodeID(v_bigger)
        n_middle = XLNodeID(val)
        n_smaller = XLNodeID(v_smaller)

        # compare them
        self.assertTrue(n_bigger > n_middle)
        self.assertTrue(n_smaller < n_middle)

        # equality checks
        self.assertEqual(n_middle, n_middle)
        self.assertFalse(n_bigger == n_middle)
        self.assertFalse(n_middle == n_smaller)

    def test_comparion(self):
        """ Test comparison operators for IDs of standard lengths. """
        for length in [SHA1_BIN_LEN, SHA2_BIN_LEN, SHA3_BIN_LEN]:
            self.do_test_comparison(length)

if __name__ == '__main__':
    unittest.main()


tests/test_xlnodeid.py
#!/usr/bin/env python3
# xlnodeid_py/test_node_id.py

""" Verify that an XLNodeID behaves like one. """

import time
import unittest
from copy import deepcopy

from rnglib import SimpleRNG
from xlattice import SHA1_BIN_LEN, SHA2_BIN_LEN, SHA3_BIN_LEN
from xlnodeid import XLNodeID, XLNodeIDError


class TestNodeID(unittest.TestCase):
    """ Verify that an XLNodeID behaves like one. """

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    def test_valid_node_id(self):
        """
        Tests the funnction of that name.
        """

        # tests that should fail
        self.assertFalse(XLNodeID.is_valid_node_id(None))
        self.assertFalse(XLNodeID.is_valid_node_id('foo'))  # not bytes-like
        self.assertFalse(XLNodeID.is_valid_node_id(b'bar'))  # wrong length
        self.assertFalse(XLNodeID.is_valid_node_id(42))     # an int

        # tests that should succeed
        val = bytes(SHA1_BIN_LEN)
        self.assertTrue(XLNodeID.is_valid_node_id(val))
        val = bytes(SHA2_BIN_LEN)
        self.assertTrue(XLNodeID.is_valid_node_id(val))
        val = bytes(SHA3_BIN_LEN)
        self.assertTrue(XLNodeID.is_valid_node_id(val))

    def expect_failure(self, val):
        """ Expect object construction to fail. """
        try:
            XLNodeID(val)
            self.fail("XLNodeID constructed with bad ID")
        except XLNodeIDError:
            pass

    def expect_success(self, val):
        """
        Try to build a NodeID with val (a bytes valaue).
        """
        try:
            nodeid = XLNodeID(val)
            # succeeded
        except XLNodeIDError:
            self.fail("ctor raised with good ID")

        val2 = nodeid.value
        self.assertIsNotNone(val2)
        self.assertFalse(val2 is val)   # not the same object
        self.assertEqual(val2, val)     # but a valid deep copy

    def test_ctor(self):
        """
        Tests the XLNodeID constructor.
        """

        # tests that should fail
        self.expect_failure(None)    # id may not be None
        self.expect_failure('foo')   # not bytes-like
        self.expect_failure(b'bar')  # wrong length
        self.expect_failure(42)      # an int

        self.expect_failure(bytes(SHA1_BIN_LEN - 1))
        self.expect_failure(bytes(SHA1_BIN_LEN + 1))
        self.expect_failure(bytes(SHA2_BIN_LEN - 1))
        self.expect_failure(bytes(SHA2_BIN_LEN + 1))
        self.expect_failure(bytes(SHA3_BIN_LEN - 1))
        self.expect_failure(bytes(SHA3_BIN_LEN + 1))

        # tests that should succeed
        val = self.rng.some_bytes(SHA1_BIN_LEN)
        self.expect_success(val)
        val = self.rng.some_bytes(SHA2_BIN_LEN)
        self.expect_success(val)
        val = self.rng.some_bytes(SHA3_BIN_LEN)
        self.expect_success(val)

    def do_test_cloning(self, length):
        """ Verify that cloning works for a given number of bytes. """
        val = self.rng.some_bytes(length)
        id1 = XLNodeID(val)
        id2 = id1.clone()
        self.assertTrue(id1 is not id2)
        self.assertEqual(id1.value, id2.value)

    def test_cloning(self):
        """ Test cloning for bytes-like objects of a given number of bytes. """
        for length in [SHA1_BIN_LEN, SHA2_BIN_LEN, SHA3_BIN_LEN]:
            self.do_test_cloning(length)

    def do_test_comparison(self, length):
        """
        For a quasi-random byte sequence of a given length, verify that
        comparison opertors work.
        """
        val = self.rng.some_bytes(length)       # a byte array of that length

        # pick a random index into that byte array
        ndx = 1 + self.rng.next_int16(length - 1)
        if val[ndx] == 0:
            val[ndx] = 1
        if val[ndx] == 255:
            val[ndx] = 254

        # make a couple of clones of the byte array
        v_bigger = deepcopy(val)
        v_bigger[ndx] += 1
        v_smaller = deepcopy(val)
        v_smaller[ndx] -= 1

        self.assertTrue(v_bigger > val)
        self.assertTrue(v_smaller < val)

        # use these values to make NodeIDs
        n_bigger = XLNodeID(v_bigger)
        n_middle = XLNodeID(val)
        n_smaller = XLNodeID(v_smaller)

        # compare them
        self.assertTrue(n_bigger > n_middle)
        self.assertTrue(n_smaller < n_middle)

        # equality checks
        self.assertEqual(n_middle, n_middle)
        self.assertFalse(n_bigger == n_middle)
        self.assertFalse(n_middle == n_smaller)

    def test_comparion(self):
        """ Test comparison operators for IDs of standard lengths. """
        for length in [SHA1_BIN_LEN, SHA2_BIN_LEN, SHA3_BIN_LEN]:
            self.do_test_comparison(length)


if __name__ == '__main__':
    unittest.main()


src/xlnodeid/__init__.py
# xlnodeid/__init__.py

""" NodeID library for python XLattice packages. """

from binascii import b2a_hex
from copy import deepcopy

from xlattice import SHA1_BIN_LEN, SHA2_BIN_LEN, SHA3_BIN_LEN

__version__ = '0.0.12'
__version_date__ = '2018-03-23'

__all__ = ['__version__', '__version_date__',
           'XLNodeIDError', 'XLNodeID']


class XLNodeIDError(RuntimeError):
    """ General purpose exception for the package. """


class XLNodeID(object):
    """ Unique identifier for an XLattice Node. """

    def __init__(self, ident):
        if not ident:
            raise XLNodeIDError("id may not be None or empty")
        if not isinstance(ident, (bytearray, bytes)):
            raise XLNodeIDError("NodeID value must be bytes-like")

        length = len(ident)
        if length != SHA1_BIN_LEN and length != SHA2_BIN_LEN and \
                length != SHA3_BIN_LEN:
            raise XLNodeIDError("invalid nodeID length %d" % length)

        # it's a valid ID, so deep copy it
        self._node_id = bytes(deepcopy(ident))

    @property
    def value(self):
        """ Return a deep copy of the underlying byte sequence. """

        return deepcopy(self._node_id)

    @staticmethod
    def is_valid_node_id(val):
        """ Return whether val is a valid XLNodeID value. """
        if not val:
            return False
        if not isinstance(val, bytes) and not isinstance(val, bytearray):
            return False

        length = len(val)
        if length != SHA1_BIN_LEN and length != SHA2_BIN_LEN and \
                length != SHA3_BIN_LEN:
            return False

        return True

    def clone(self):
        """ Return a deep copy of the XLNodeID instance. """
        return XLNodeID(self._node_id)

    def __eq__(self, other):
        if not isinstance(other, XLNodeID):
            return False
        return self._node_id == other.value

    def __lt__(self, other):
        return self._node_id < other.value

    def __str__(self):
        return b2a_hex(self._node_id)

    def __len__(self):
        return len(self._node_id)
